File: app/src/main/java/com/project/scratchstudio/kith_andoid/UI/BoardInfo/BoardInfoFragment.java
package com.project.scratchstudio.kith_andoid.UI.BoardInfo;

import android.content.Intent;
import android.icu.text.DateFormat;
import android.icu.text.SimpleDateFormat;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.project.scratchstudio.kith_andoid.Activities.HomeActivity;
import com.project.scratchstudio.kith_andoid.Activities.ProfileActivity;
import com.project.scratchstudio.kith_andoid.R;
import com.project.scratchstudio.kith_andoid.UI.Comments.CommentListFragment;
import com.project.scratchstudio.kith_andoid.UI.NewEditBoard.NewEditBoardFragment;
import com.project.scratchstudio.kith_andoid.app.BaseFragment;
import com.project.scratchstudio.kith_andoid.app.FragmentType;
import com.project.scratchstudio.kith_andoid.network.ApiClient;
import com.project.scratchstudio.kith_andoid.network.apiService.BoardApi;
import com.project.scratchstudio.kith_andoid.network.apiService.UserApi;
import com.project.scratchstudio.kith_andoid.network.model.board.Board;
import com.project.scratchstudio.kith_andoid.network.model.favorite.FavoriteResponse;
import com.project.scratchstudio.kith_andoid.network.model.user.UserResponse;
import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.Picasso;

import java.util.Date;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public class BoardInfoFragment extends BaseFragment {

    private Bundle bundle;
    private Board info;
    private BoardApi boardApi;
    private UserApi userApi;
    private CompositeDisposable disposable = new CompositeDisposable();

    private ImageView photo;

    private void setIsJoin() {
        CheckBox favorite = getActivity().findViewById(R.id.heart);
        favorite.setChecked(true);
        favorite.setEnabled(true);
    }

    public static BoardInfoFragment newInstance(Bundle bundle) {
        BoardInfoFragment fragment = new BoardInfoFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        bundle = getArguments();
        bundle.putBoolean("is_edit", false);
        if (bundle != null) {
            info = (Board) bundle.getSerializable("board");
        }
        return inflater.inflate(R.layout.fragment_announcement_information, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        setButtonsListener();

        TextView title = view.findViewById(R.id.title);
        TextView description = view.findViewById(R.id.description);
        TextView owner = view.findViewById(R.id.owner);
        TextView phone = view.findViewById(R.id.phone);
        TextView creationDate = view.findViewById(R.id.creationDate);
        photo = view.findViewById(R.id.photo);

        title.setText(info.title);

        try {
            DateFormat inputFormat;
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
                inputFormat = new SimpleDateFormat("yyyy-MM-dd");
                DateFormat outputFormat = new SimpleDateFormat("dd.MM.yyyy");
                Date newCreationDateFormat = null;
                if (info.startDate.equals("null")) {
                    creationDate.setVisibility(View.GONE);
                } else {
                    newCreationDateFormat = inputFormat.parse(info.startDate.replaceAll("\\s.*$", ""));
                }
                creationDate.setText(getString(R.string.create_date, outputFormat.format(newCreationDateFormat)));
            } else {
                String errDate = "Создано: " + info.startDate.replaceAll("\\s.*$", "");
                creationDate.setText(errDate);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        String ownerName = "Владелец: ";
        if (info.organizerLastName != null && !info.organizerLastName.equals("null")) {
            ownerName += info.organizerLastName + " ";
        } else {
            owner.setVisibility(View.GONE);
        }
        if (!info.organizerName.equals("null")) {
            owner.setVisibility(View.VISIBLE);
            ownerName += info.organizerName;
        }
        owner.setText(ownerName);

        if (info.organizerPhone != null && !info.organizerPhone.equals("null")) {
            phone.setText(getString(R.string.organizer_phone, info.organizerPhone));
        } else {
            phone.setVisibility(View.GONE);
        }

        description.setText(info.description);
        if (info.url != null && !info.url.equals("null") && !info.url.equals("")) {
            Picasso.with(getActivity()).load(info.url.replaceAll("@[0-9]*", ""))
                    .error(R.drawable.newspaper)
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(photo);
            photo.setScaleType(ImageView.ScaleType.CENTER_CROP);
        }

        if (info.organizerId == HomeActivity.getMainUser().getId()) {
            ImageButton button = getActivity().findViewById(R.id.edit);
            button.setVisibility(View.VISIBLE);
            button.setOnClickListener(this::onClickEdit);
        }

        if (info.subscriptionOnBoard == 1) {
            setIsJoin();
        }

        userApi = ApiClient.getClient(getContext()).create(UserApi.class);
        boardApi = ApiClient.getClient(getContext()).create(BoardApi.class);
    }

    public void reloadPhoto() {
        if (info.url != null && !info.url.equals("null") && !info.url.equals("")) {
            Picasso.with(getActivity()).load(info.url.replaceAll("@[0-9]*", ""))
                    .error(R.drawable.newspaper)
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(photo);
            photo.setScaleType(ImageView.ScaleType.CENTER_CROP);
        }
    }

    private void setButtonsListener() {
        ImageButton back = getActivity().findViewById(R.id.back);
        back.setOnClickListener(this::onClickBack);
        CheckBox favorite = getActivity().findViewById(R.id.heart);
        favorite.setOnClickListener(this::onClickJoin);
        Button comments = getActivity().findViewById(R.id.comments);
        comments.setOnClickListener(this::onClickComments);
        Button profile = getActivity().findViewById(R.id.user_profile);
        profile.setOnClickListener(this::onClickProfile);
        TextView phone = getActivity().findViewById(R.id.phone);
        phone.setOnClickListener(this::onClickPhone);
    }

    private void onClickPhone(View view) {
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setData(Uri.parse("tel:" + info.organizerPhone));
        startActivity(intent);
    }

    private void onClickProfile(View view) {
        view.setEnabled(false);
        disposable.add(
                userApi.getUser(info.organizerId)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<UserResponse>() {
                            @Override
                            public void onSuccess(UserResponse response) {
                                if (response.getStatus()) {
                                    Log.i("Get User Resp", response.toString());
                                    response.getUser().setId(info.organizerId);
                                    if (response.getUser().photo != null) {
                                        response.getUser().setUrl(response.getUser().photo.replaceAll("\\/", "/"));
                                    }
                                    Bundle bundle = new Bundle();
                                    bundle.putBoolean("another_user", true);
                                    Intent intent = new Intent(getContext(), ProfileActivity.class);
                                    intent.putExtra("user", response.getUser());

                                    if (response.getUser().id != HomeActivity.getMainUser().id) {
                                        intent.putExtra("another_user", true);
                                    } else {
                                        intent.putExtra("another_user", false);
                                    }

                                    startActivity(intent);
                                } else {
                                    Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                    view.setEnabled(true);
                                }
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("BoardFragmentInfo", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                view.setEnabled(true);
                            }
                        })
        );
    }

    private void onClickEdit(View view) {
        bundle.putBoolean("is_edit", true);
        bundle.putSerializable("board", info);
        ((HomeActivity) getActivity()).addFragment(NewEditBoardFragment.newInstance(bundle), FragmentType.BOARD_NEW_EDIT.name());
    }

    public void onClickBack(View view) {
        ((HomeActivity) getActivity()).backFragment();
    }

    private void onClickComments(View view) {
        bundle.putInt("board_id", info.id);
        bundle.putString("board_title", info.title);
        ((HomeActivity) getContext()).addFragment(CommentListFragment.newInstance(bundle), FragmentType.COMMENT_LIST.name());
    }

    private void onClickJoin(View view) {
        view.setEnabled(false);
        if (info.subscriptionOnBoard != 1) {
            subscribeAnnouncement(HomeActivity.getMainUser().getId(), info, (CheckBox) view);
        } else {
            unsubscribeAnnouncement(HomeActivity.getMainUser().getId(), info, (CheckBox) view);
        }
    }

    private void subscribeAnnouncement(int user_id, Board board, CheckBox favorite) {
        disposable.add(
                boardApi.subscribeAnnouncement(String.valueOf(user_id), String.valueOf(board.id))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<FavoriteResponse>() {
                            @Override
                            public void onSuccess(FavoriteResponse response) {
                                if (response.getStatus()) {
                                    board.subscriptionOnBoard = 1;
                                    Toast.makeText(getContext(), "Добавлено в избранное", Toast.LENGTH_SHORT).show();
                                } else {
                                    Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                    favorite.setChecked(!favorite.isChecked());
                                }
                                favorite.setEnabled(true);
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("BoardsFragment", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                favorite.setChecked(!favorite.isChecked());
                                favorite.setEnabled(true);
                            }
                        })
        );
    }

    private void unsubscribeAnnouncement(int user_id, Board board, CheckBox favorite) {
        disposable.add(
                boardApi.unsubscribeAnnouncement(String.valueOf(user_id), String.valueOf(board.id))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<FavoriteResponse>() {
                            @Override
                            public void onSuccess(FavoriteResponse response) {
                                if (response.getStatus()) {
                                    board.subscriptionOnBoard = 0;
                                    Toast.makeText(getContext(), "Удалено из избранного", Toast.LENGTH_SHORT).show();
                                } else {
                                    Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                    favorite.setChecked(!favorite.isChecked());
                                }
                                favorite.setEnabled(true);
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("BoardsFragment", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                favorite.setChecked(!favorite.isChecked());
                                favorite.setEnabled(true);
                            }
                        })
        );
    }

    @Override
    public boolean onBackPressed() {
        return ((HomeActivity) getActivity()).backFragment();
    }
}

File: app/src/main/java/com/project/scratchstudio/kith_andoid/ui/home_package/BoardInfo/BoardInfoFragment.kt
package com.project.scratchstudio.kith_andoid.ui.home_package.BoardInfo

import android.accounts.NetworkErrorException
import android.content.Intent
import android.icu.text.DateFormat
import android.icu.text.SimpleDateFormat
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.CheckBox
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.project.scratchstudio.kith_andoid.Activities.HomeActivity
import com.project.scratchstudio.kith_andoid.R
import com.project.scratchstudio.kith_andoid.UserPresenter
import com.project.scratchstudio.kith_andoid.app.BaseFragment
import com.project.scratchstudio.kith_andoid.app.FragmentType
import com.project.scratchstudio.kith_andoid.network.ApiClient
import com.project.scratchstudio.kith_andoid.network.apiService.BoardApi
import com.project.scratchstudio.kith_andoid.network.apiService.UserApi
import com.project.scratchstudio.kith_andoid.network.model.board.Board
import com.project.scratchstudio.kith_andoid.network.model.favorite.FavoriteResponse
import com.project.scratchstudio.kith_andoid.network.model.user.UserResponse
import com.project.scratchstudio.kith_andoid.ui.home_package.Comments.CommentListFragment
import com.project.scratchstudio.kith_andoid.ui.home_package.neweditboard.NewEditBoardFragment
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.Picasso
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import java.util.Date

class BoardInfoFragment : BaseFragment() {

    private var bundle: Bundle? = null
    private var info: Board? = null
    private var boardApi: BoardApi? = null
    private var userApi: UserApi? = null
    private val disposable = CompositeDisposable()
    private var userPresenter: UserPresenter? = null

    private var photo: ImageView? = null

    private fun setIsJoin() {
        val favorite = activity!!.findViewById<CheckBox>(R.id.heart)
        favorite.isChecked = true
        favorite.isEnabled = true
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        bundle = arguments
        bundle!!.putBoolean("is_edit", false)
        if (bundle != null) {
            info = bundle!!.getSerializable("board") as Board
        }
        return inflater.inflate(R.layout.fragment_announcement_information, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        userPresenter = UserPresenter(context!!)
        setButtonsListener()

        val title = view.findViewById<TextView>(R.id.title)
        val description = view.findViewById<TextView>(R.id.description)
        val owner = view.findViewById<TextView>(R.id.owner)
        val phone = view.findViewById<TextView>(R.id.phone)
        val creationDate = view.findViewById<TextView>(R.id.creationDate)
        photo = view.findViewById(R.id.photo)

        title.text = info!!.title

        try {
            val inputFormat: DateFormat
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
                inputFormat = SimpleDateFormat("yyyy-MM-dd")
                val outputFormat = SimpleDateFormat("dd.MM.yyyy")
                var newCreationDateFormat: Date? = null
                if (info!!.startDate == "null") {
                    creationDate.visibility = View.GONE
                } else {
                    newCreationDateFormat = inputFormat.parse(info!!.startDate!!.replace("\\s.*$".toRegex(), ""))
                }
                creationDate.text = getString(R.string.create_date, outputFormat.format(newCreationDateFormat))
            } else {
                val errDate = "Создано: " + info!!.startDate!!.replace("\\s.*$".toRegex(), "")
                creationDate.text = errDate
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }

        val ownerName = StringBuilder()
        if (info!!.organizerName != null && info!!.organizerName != "null") {
            ownerName.append(info!!.organizerName)
        } else {
            if (info!!.organizerLastName != null && info!!.organizerLastName != "null") {
                ownerName.append(info!!.organizerLastName)
                owner.append(" ")
                if (info!!.organizerFirstName != null && info!!.organizerFirstName != "null") {
                    owner.append(info!!.organizerFirstName)
                }
            } else {
                owner.visibility = View.GONE
            }
        }
        owner.text = getString(R.string.board_info_owner, ownerName)

        if (info!!.organizerPhone != null && info!!.organizerPhone != "null") {
            phone.text = getString(R.string.organizer_phone, info!!.organizerPhone)
        } else {
            phone.visibility = View.GONE
        }

        description.text = info!!.description
        if (info!!.url != null && info!!.url != "null" && info!!.url != "") {
            Picasso.with(activity).load(info!!.url!!.replace("@[0-9]*".toRegex(), ""))
                    .error(R.drawable.newspaper)
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(photo)
            photo!!.scaleType = ImageView.ScaleType.CENTER_CROP
        }

        if (info!!.organizerId == HomeActivity.mainUser!!.id) {
            val button = activity!!.findViewById<ImageButton>(R.id.edit)
            button.visibility = View.VISIBLE
            button.setOnClickListener { this.onClickEdit(it) }
        }

        if (info!!.subscriptionOnBoard == 1) {
            setIsJoin()
        }

        userApi = ApiClient.getClient(context!!).create<UserApi>(UserApi::class.java!!)
        boardApi = ApiClient.getClient(context!!).create<BoardApi>(BoardApi::class.java!!)
    }

    fun reloadPhoto() {
        if (info!!.url != "" && info!!.url != "null") {
            Picasso.with(activity).load(info!!.url!!.replace("@[0-9]*".toRegex(), ""))
                    .error(R.drawable.newspaper)
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(photo)
            photo!!.scaleType = ImageView.ScaleType.CENTER_CROP
        }
    }

    private fun setButtonsListener() {
        val back = activity!!.findViewById<ImageButton>(R.id.back)
        back.setOnClickListener { this.onClickBack(it) }
        val favorite = activity!!.findViewById<CheckBox>(R.id.heart)
        favorite.setOnClickListener { this.onClickJoin(it) }
        val comments = activity!!.findViewById<Button>(R.id.comments)
        comments.setOnClickListener { this.onClickComments(it) }
        val profile = activity!!.findViewById<Button>(R.id.user_profile)
        profile.setOnClickListener { this.onClickProfile(it) }
        val phone = activity!!.findViewById<TextView>(R.id.phone)
        phone.setOnClickListener { this.onClickPhone(it) }
    }

    private fun onClickPhone(view: View) {
        val intent = Intent(Intent.ACTION_DIAL)
        intent.data = Uri.parse("tel:" + info!!.organizerPhone!!)
        startActivity(intent)
    }

    private fun onClickProfile(view: View) {
        view.isEnabled = false
        userPresenter!!.getUser(object : UserPresenter.GetUserCallback {
            override fun onSuccess(userResponse: UserResponse) {
                if (userResponse.status) {
                    Log.i("Get User Resp", userResponse.toString())
                    userResponse.user!!.id = info!!.organizerId
                    if (userResponse.user!!.photo != "") {
                        userResponse.user!!.photo = userResponse.user!!.photo!!.replace("\\/".toRegex(), "/")
                    }
                    userPresenter!!.openProfile(userResponse.user!!)
                } else {
                    Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                }
                view.isEnabled = true
            }

            override fun onError(networkError: NetworkErrorException) {
                Log.e("BoardFragmentInfo", "onError: " + networkError.message)
                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                view.isEnabled = true
            }
        }, info!!.organizerId)
    }

    private fun onClickEdit(view: View) {
        bundle!!.putBoolean("is_edit", true)
        bundle!!.putSerializable("board", info)
        (activity as HomeActivity).addFragment(NewEditBoardFragment.newInstance(bundle!!), FragmentType.BOARD_NEW_EDIT.name)
    }

    fun onClickBack(view: View) {
        (activity as HomeActivity).backFragment()
    }

    private fun onClickComments(view: View) {
        bundle!!.putInt("board_id", info!!.id)
        bundle!!.putString("board_title", info!!.title)
        (context as HomeActivity).addFragment(CommentListFragment.newInstance(bundle!!), FragmentType.COMMENT_LIST.name)
    }

    private fun onClickJoin(view: View) {
        view.isEnabled = false
        if (info!!.subscriptionOnBoard != 1) {
            subscribeAnnouncement(HomeActivity.mainUser!!.id, info!!, view as CheckBox)
        } else {
            unsubscribeAnnouncement(HomeActivity.mainUser!!.id, info!!, view as CheckBox)
        }
    }

    private fun subscribeAnnouncement(user_id: Int, board: Board, favorite: CheckBox) {
        disposable.add(
                boardApi!!.subscribeAnnouncement(user_id.toString(), board.id.toString())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<FavoriteResponse>() {
                            override fun onSuccess(response: FavoriteResponse) {
                                if (response.status) {
                                    board.subscriptionOnBoard = 1
                                    Toast.makeText(context, "Добавлено в избранное", Toast.LENGTH_SHORT).show()
                                } else {
                                    Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                    favorite.isChecked = !favorite.isChecked
                                }
                                favorite.isEnabled = true
                            }

                            override fun onError(e: Throwable) {
                                Log.e("BoardsFragment", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                favorite.isChecked = !favorite.isChecked
                                favorite.isEnabled = true
                            }
                        })
        )
    }

    private fun unsubscribeAnnouncement(user_id: Int, board: Board, favorite: CheckBox) {
        disposable.add(
                boardApi!!.unsubscribeAnnouncement(user_id.toString(), board.id.toString())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<FavoriteResponse>() {
                            override fun onSuccess(response: FavoriteResponse) {
                                if (response.status) {
                                    board.subscriptionOnBoard = 0
                                    Toast.makeText(context, "Удалено из избранного", Toast.LENGTH_SHORT).show()
                                } else {
                                    Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                    favorite.isChecked = !favorite.isChecked
                                }
                                favorite.isEnabled = true
                            }

                            override fun onError(e: Throwable) {
                                Log.e("BoardsFragment", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                favorite.isChecked = !favorite.isChecked
                                favorite.isEnabled = true
                            }
                        })
        )
    }

    override fun onBackPressed(): Boolean {
        return (activity as HomeActivity).backFragment()
    }

    companion object {

        fun newInstance(bundle: Bundle): BoardInfoFragment {
            val fragment = BoardInfoFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
